src/Terminal/commands/grep.ts

```diff
@@ -0,0 +1,445 @@
+    [context, this.args] = this.spliceOptParam(VALID_ARGS.isPreContext);
+    if (!context) [context, this.args] = this.spliceOptParam(VALID_ARGS.isContext);
+    if (!context) [context, this.args] = this.spliceOptParam(VALID_ARGS.isPostContext);
+
+    const [options, otherArgs] = this.args.reduce(
+      ([options, otherArgs]: [Options, string[]], fullArg: string): [Options, string[]] => {
+        let isOption = false;
+        [options, isOption] = this.mapArgToOpts(fullArg, options);
+        return isOption ? [options, otherArgs] : [options, [...otherArgs, fullArg]];
+      },
+      [this.initOptions, []],
+    );
+    const outFileStr = outFile ?? "";
+    const limitNum = limit ?? "";
+    const contextNum = context ?? "";
+
+    return [options, otherArgs, outFileStr, contextNum, limitNum];
+  }
+
+  spliceOptParam(validArgs: ArgStrings): [string, string[]] | [undefined, string[]] {
+    const argIndex = [...validArgs.long, ...validArgs.short].reduce((ret: number, arg: string) => {
+      const argIndex = this.args.indexOf(arg);
+      return argIndex > -1 ? argIndex : ret;
+    }, NaN);
+
+    if (isNaN(argIndex)) return [undefined, this.args];
+
+    const nextArg = this.args.splice(argIndex + 1, 1)[0];
+
+    return [nextArg, this.args];
+  }
+}
+
+class Results {
+  lines: Line[];
+  areEdited: boolean;
+  numMatches: number;
+  options: Options;
+  matchCounter: number;
+  matchLimit: number;
+
+  constructor(results: Line[], options: Options, matchLimit: number) {
+    this.lines = results;
+    this.options = options;
+    this.areEdited = results.some((line) => line.isMatched);
+    this.numMatches = results.reduce((acc, result) => acc + Number(result.isMatched), 0);
+    this.matchLimit = matchLimit;
+    this.matchCounter = 0;
+  }
+
+  addContext(context: number): Results {
+    const nContext = isNaN(Number(context)) ? 0 : Number(context);
+    for (const [editLineIndex, line] of this.lines.entries()) {
+      if (!line.isMatched) continue;
+      for (let contextLineIndex = 0; contextLineIndex <= nContext; contextLineIndex++) {
+        let contextLine;
+        if (this.options.isPreContext) contextLine = this.lines[editLineIndex - contextLineIndex];
+        else if (this.options.isPostContext) contextLine = this.lines[editLineIndex + contextLineIndex];
+        else if (this.options.isContext)
+          contextLine = this.lines[editLineIndex - Math.floor(nContext / 2) + contextLineIndex];
+        else contextLine = line;
+
+        if (contextLine && !line.isFileSep && line.filename === contextLine.filename) contextLine.isPrint = true;
+      }
+    }
+    return this;
+  }
+
+  splitAndFilter(): [string[], string[]] {
+    return this.lines.reduce(
+      ([rawResult, prettyResult]: [string[], string[]], lineInfo: Line): [string[], string[]] => {
+        return lineInfo.isPrint === this.options.isInvertMatch
+          ? [rawResult, prettyResult]
+          : [
+              [...rawResult, lineInfo.lines.rawLine],
```

Comment:
I can't bite my tongue about this. XD Each iteration, you are destructuring the previous array into a new array, and then adding a new item at the end. In other words, you are turning what should be a simple O(N) operation into O(N^2).

In addition, you have this complicated ternary just to create a bypass for the "don't do anything case." This whole thing becomes *much* simpler if you just do it iteratively:
```suggestion
  if (lineInfo.isPrint !== this.options.isInvertMatch) return [[], []];
  const rawResult = [], prettyResult = [];
  for (const lineInfo of this.lines) {
    rawResult.push(lineInfo.lines.rawLine);
    prettyResult.push(lineInfo.lines.prettyLine);
  }
  return [rawResult, prettyResult];
```

Reply:
No tongue biting necessary. :). I am still quite blind to these things and I'm very glad you point them out!

:+1: